File: src/main/javascript/components/specifications/ContactWasMember.ts
/// <reference path="../../types/app.ts" />

module CSRS {
    export class ContactWasMember implements Specification<Contact> {
        constructor (
            private yearsRequired: Array<number | string>,
            private yearsForbidden: Array<number | string>
        ) {}

        toPredicate () : Predicate<Contact> {
            return null;
        }

        toJSON () : any {
            return ({
                '@type': 'ContactWasMember', 
                yearsRequired: this.yearsRequired,
                yearsForbidden: this.yearsForbidden,
            });
        }
    }

    angular.module('csrsApp').factory("ContactWasMember", () => {
        return ContactWasMember;
    });
}

File: src/main/java/com/fulliautomatix/csrs/specification/ContactWasMember.java
package com.fulliautomatix.csrs.specification;

import org.springframework.data.jpa.domain.Specification;

import com.fasterxml.jackson.annotation.*;
import com.voodoodyne.jackson.jsog.JSOGGenerator;

import com.fulliautomatix.csrs.domain.Contact;
import com.fulliautomatix.csrs.domain.Contact_;

import javax.persistence.criteria.*;
import java.util.*;

@lombok.Data
@lombok.NoArgsConstructor @lombok.AllArgsConstructor
@lombok.EqualsAndHashCode(callSuper=false)
@JsonTypeName("ContactWasMember")
public class ContactWasMember extends Spec<Contact> {
    @lombok.Getter @lombok.Setter
    private Set<Integer> yearsRequired;

    @lombok.Getter @lombok.Setter
    private Set<Integer> yearsForbidden;

    public ContactWasMember (Integer[] yr, Integer[] yf) {
        this(
            yr == null ? (Set<Integer>) null : new HashSet<Integer>(Arrays.asList(yr)), 
            yf == null ? (Set<Integer>) null : new HashSet<Integer>(Arrays.asList(yf))
        );
    }

    @Override
    public Predicate toPredicate (Root<Contact> contact, CriteriaQuery<?> query, CriteriaBuilder cb) {
        Predicate positive;
        
        if ((yearsRequired == null) || (yearsRequired.size() == 0)) {
            // No yearsRequired, so the positive aspect is ContactWasMember;
            positive = new ContactWasEverMember().toPredicate(contact, query, cb);
        } else {
            positive = new ContactWasMemberInAll(yearsRequired).toPredicate(contact, query, cb);
        }

        if ((yearsForbidden == null) || (yearsForbidden.size() == 0)) {
            // If nothing forbidden, just return the positive part
            return positive;
        } else {
            return cb.and(
                positive,
                new ContactWasMemberInNone(yearsForbidden).toPredicate(contact, query, cb)
            );
        }
    }
}
